controllers/openstackcluster_controller.go

```diff
@@ -627,3 +617,51 @@ func handleUpdateOSCError(openstackCluster *infrav1.OpenStackCluster, message er
 	openstackCluster.Status.FailureReason = &err
 	openstackCluster.Status.FailureMessage = pointer.String(message.Error())
 }
+
+// filterSubnets retrieves the subnets based on the Subnet filters specified on OpenstackCluster.
+func filterSubnets(networkingService *networking.Service, openStackCluster *infrav1.OpenStackCluster) ([]infrav1.Subnet, error) {
+	var subnets []infrav1.Subnet
+	openStackClusterSubnets := openStackCluster.Spec.Subnets
+	if len(openStackClusterSubnets) == 0 {
+		empty := &infrav1.SubnetFilter{}
+		listOpt := empty.ToListOpt()
+		listOpt.NetworkID = openStackCluster.Status.Network.ID
+		filteredSubnets, err := networkingService.GetSubnetsByFilter(listOpt)
+		if err != nil {
+			err = fmt.Errorf("failed to find subnets: %w", err)
+			if errors.Is(err, networking.ErrFilterMatch) {
+				handleUpdateOSCError(openStackCluster, err)
+			}
+			return nil, err
+		}
+		if len(filteredSubnets) > 2 {
+			return nil, fmt.Errorf("more than two subnets found in the Network. Specify the subnets in the OpenStackCluster.Spec instead")
+		}
+		for i := range filteredSubnets {
```

Comment:
I'm not sure why not `for _, subnet := range filteredSubnets` so that we can use `subnet` directly later on.